{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f382f0a",
   "metadata": {},
   "source": [
    "### Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e13de0f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re  \n",
    "import fnmatch \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n",
    "\n",
    "from sklearn.ensemble import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "063e20b4",
   "metadata": {},
   "source": [
    "### Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9e1a46a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "parquet_dir = './Data/parquet/'\n",
    "output_file = './Data/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f27bf68",
   "metadata": {},
   "source": [
    "### Read Population Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3c996649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123417</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00276</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123424</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00288</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME1839246</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample name     Sex Biosample ID Population code Population name  \\\n",
       "0     HG00271    male   SAME123417             FIN         Finnish   \n",
       "1     HG00276  female   SAME123424             FIN         Finnish   \n",
       "2     HG00288  female  SAME1839246             FIN         Finnish   \n",
       "\n",
       "  Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "0                  EUR    European Ancestry                   FIN   \n",
       "1                  EUR    European Ancestry                   FIN   \n",
       "2                  EUR    European Ancestry                   FIN   \n",
       "\n",
       "                                    Data collections  \n",
       "0  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "1  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_labels = pd.read_csv(parquet_dir + 'igsr_samples.tsv', sep='\\t')\n",
    "pop_labels.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35151422",
   "metadata": {},
   "source": [
    "### Read, Clean, and Merge SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa586ae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_and_clean_chromosome_files(folder_path):\n",
    "\n",
    "    for chr_num in range(1, 23):\n",
    "        \n",
    "        file_path = f\"{folder_path}/chr{chr_num}.parquet\"\n",
    "        df = pd.read_parquet(file_path, engine='pyarrow')\n",
    "        \n",
    "        missing_percentage = df.isnull().mean() * 100\n",
    "        cols_to_keep = missing_percentage[missing_percentage <= 5].index\n",
    "        cleaned_df = df[cols_to_keep]\n",
    "        \n",
    "        if chr_num == 1:\n",
    "            final = cleaned_df.copy()\n",
    "        else:\n",
    "            final = final.merge(cleaned_df, on='Person_ID')\n",
    "                \n",
    "    return final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "09a60b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = read_and_clean_chromosome_files(parquet_dir)\n",
    "result_df = result_df.set_index('Person_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "eceb4311",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs3094315</th>\n",
       "      <th>rs3131972</th>\n",
       "      <th>rs12124819</th>\n",
       "      <th>rs11240777</th>\n",
       "      <th>rs6681049</th>\n",
       "      <th>rs4970383</th>\n",
       "      <th>rs4475691</th>\n",
       "      <th>rs7537756</th>\n",
       "      <th>rs13302982</th>\n",
       "      <th>rs1110052</th>\n",
       "      <th>...</th>\n",
       "      <th>rs10451</th>\n",
       "      <th>rs715586</th>\n",
       "      <th>rs8137951</th>\n",
       "      <th>rs2301584</th>\n",
       "      <th>rs756638</th>\n",
       "      <th>rs3810648</th>\n",
       "      <th>rs2285395</th>\n",
       "      <th>rs13056621</th>\n",
       "      <th>rs5771007</th>\n",
       "      <th>rs3888396</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Person_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 909218 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rs3094315  rs3131972  rs12124819  rs11240777  rs6681049  rs4970383  \\\n",
       "Person_ID                                                                       \n",
       "HG00096            1          1           0           1          2          0   \n",
       "HG00097            2          2           0           0          2          1   \n",
       "HG00099            1          1           0           1          2          0   \n",
       "\n",
       "           rs4475691  rs7537756  rs13302982  rs1110052  ...  rs10451  \\\n",
       "Person_ID                                               ...            \n",
       "HG00096            0          0           2          1  ...        0   \n",
       "HG00097            1          1           2          0  ...        0   \n",
       "HG00099            0          0           2          2  ...        1   \n",
       "\n",
       "           rs715586  rs8137951  rs2301584  rs756638  rs3810648  rs2285395  \\\n",
       "Person_ID                                                                   \n",
       "HG00096           1          0          0         0          0          0   \n",
       "HG00097           0          0          0         0          0          0   \n",
       "HG00099           0          1          0         0          0          0   \n",
       "\n",
       "           rs13056621  rs5771007  rs3888396  \n",
       "Person_ID                                    \n",
       "HG00096             1          0          1  \n",
       "HG00097             1          0          0  \n",
       "HG00099             2          0          0  \n",
       "\n",
       "[3 rows x 909218 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8a9a2c",
   "metadata": {},
   "source": [
    "### Outlier Detection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c3f3949",
   "metadata": {},
   "source": [
    "Identify all values that are outside the range of 0 to 2 (inclusive), and set those values to NaN, which represents missing data in pandas DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b14555e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_snp_data(df):\n",
    "    \n",
    "    mask = (df < 0) | (df > 2)\n",
    "    df[mask] = np.nan\n",
    "    \n",
    "    return df\n",
    "\n",
    "result_df = clean_snp_data(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ceb30e",
   "metadata": {},
   "source": [
    "### Remove SNPs with Missing Values > 5% "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc207cd6",
   "metadata": {},
   "source": [
    "Remove: If a sample (row) or SNP (column) has a high percentage of missing values, you might consider removing it entirely from the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a492bf22",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 0.05 * result_df.shape[0]  \n",
    "result_df = result_df.dropna(axis='columns', thresh=threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8318f217",
   "metadata": {},
   "source": [
    "### Impute Missing Values with Mode"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7112e43",
   "metadata": {},
   "source": [
    "Impute: For SNPs with a small percentage of missing data, you can impute missing values. Common strategies include replacing missing values with the median or mode. In SNP data, using the most common genotype (0, 1, or 2) among the observed values for that SNP can be a sensible approach."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7460e1ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def selective_imputation(df):\n",
    "    \n",
    "    missing_columns = df.columns[df.isnull().any()].tolist()\n",
    "\n",
    "    imputer = SimpleImputer(strategy='most_frequent')\n",
    "    \n",
    "    if missing_columns:\n",
    "        df[missing_columns] = imputer.fit_transform(df[missing_columns])\n",
    "    \n",
    "    return df\n",
    "\n",
    "result_df = selective_imputation(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b236c0c",
   "metadata": {},
   "source": [
    "### Minor Allele Frequency (MAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61c7ea58",
   "metadata": {},
   "source": [
    "SNPs with a very low MAF are often excluded since they may not contribute much information for ancestry prediction and can lead to model overfitting. A common threshold is to exclude SNPs with a MAF less than 5%."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bd9988a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_maf_A(df, maf_threshold=0.05):\n",
    "\n",
    "    maf_dict = {}\n",
    "    total_alleles = 2 * len(df)\n",
    "    \n",
    "    for column in df.columns:\n",
    "        \n",
    "        counts = df[column].value_counts()\n",
    "        \n",
    "        freq_minor_allele = min((2 * counts.get(2, 0) + counts.get(1, 0)) / total_alleles,\n",
    "                                (2 * counts.get(0, 0) + counts.get(1, 0)) / total_alleles)\n",
    "        \n",
    "        maf_dict[column] = freq_minor_allele\n",
    "    \n",
    "    columns_to_keep = [column for column, maf in maf_dict.items() if maf >= maf_threshold]\n",
    "    \n",
    "    filtered_df = df[columns_to_keep]\n",
    "    \n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f3b19a2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_maf_B(dataframe, maf_threshold=0.05):\n",
    "\n",
    "    maf_dict = {}\n",
    "    to_drop  = []\n",
    "\n",
    "    for snp in dataframe.columns:\n",
    "\n",
    "        genotype_counts = dataframe[snp].value_counts()\n",
    "        \n",
    "        n_AA = genotype_counts.get(0, 0)\n",
    "        n_Aa = genotype_counts.get(1, 0)\n",
    "        n_aa = genotype_counts.get(2, 0)\n",
    "        \n",
    "        total_alleles = 2 * (n_AA + n_Aa + n_aa)\n",
    "        freq_A = (2 * n_AA + n_Aa) / total_alleles\n",
    "        freq_a = 1 - freq_A  \n",
    "        \n",
    "        maf = min(freq_A, freq_a)\n",
    "        maf_dict[snp] = maf\n",
    "        \n",
    "        if maf < maf_threshold:\n",
    "            to_drop.append(snp)\n",
    "    \n",
    "    filtered_df = dataframe.drop(columns=to_drop)\n",
    "\n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "512d0566",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 823111)\n"
     ]
    }
   ],
   "source": [
    "result_df = filter_snps_by_maf_B(result_df, maf_threshold=0.05)\n",
    "print(result_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e47cd792",
   "metadata": {},
   "source": [
    "### Hardy-Weinberg Equilibrium (HWE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9255ee00",
   "metadata": {},
   "source": [
    "Deviations from HWE can indicate problems such as genotyping errors. While checking for HWE is more common in association studies, you might decide to perform this check and exclude SNPs not in equilibrium if you suspect data quality issues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cb036515",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_hwe_A(df, hwe_p_threshold=0.01):\n",
    "\n",
    "    columns_to_keep = []\n",
    "    \n",
    "    for column in df.columns:\n",
    "        \n",
    "        genotype_counts = df[column].value_counts().sort_index()\n",
    "\n",
    "        for genotype in [0, 1, 2]:  \n",
    "            if genotype not in genotype_counts.index:\n",
    "                genotype_counts.at[genotype] = 0\n",
    "\n",
    "        p = (2 * genotype_counts[0] + genotype_counts[1]) / (2 * sum(genotype_counts.values)) \n",
    "        q = 1 - p \n",
    "\n",
    "        expected_counts = pd.Series({0: (p ** 2)  * sum(genotype_counts.values),  \n",
    "                                     1: 2 * p * q * sum(genotype_counts.values), \n",
    "                                     2: (q ** 2)  * sum(genotype_counts.values)})\n",
    "\n",
    "        chi2, p_value = stats.chisquare(f_obs=genotype_counts, f_exp=expected_counts)\n",
    "\n",
    "        if p_value >= hwe_p_threshold:\n",
    "            columns_to_keep.append(column)\n",
    "            \n",
    "    filtered_df = df[columns_to_keep]\n",
    "    \n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a3bf51fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_hwe_B(dataframe, alpha=0.01):\n",
    "\n",
    "    hwe_results = {}\n",
    "    \n",
    "    for snp in dataframe.columns:\n",
    "        \n",
    "        genotype_counts = dataframe[snp].value_counts()\n",
    "        \n",
    "        obs_AA = genotype_counts.get(0, 0)\n",
    "        obs_Aa = genotype_counts.get(1, 0)\n",
    "        obs_aa = genotype_counts.get(2, 0)\n",
    "        \n",
    "        total_alleles = 2 * (obs_AA + obs_Aa + obs_aa)\n",
    "        p = (2 * obs_AA + obs_Aa) / total_alleles \n",
    "        q = 1 - p  \n",
    "        \n",
    "        exp_AA = (p ** 2)  * (obs_AA + obs_Aa + obs_aa)\n",
    "        exp_Aa = 2 * p * q * (obs_AA + obs_Aa + obs_aa)\n",
    "        exp_aa = (q ** 2)  * (obs_AA + obs_Aa + obs_aa)\n",
    "        \n",
    "        observed = np.array([obs_AA, obs_Aa, obs_aa])\n",
    "        expected = np.array([exp_AA, exp_Aa, exp_aa])\n",
    "        chi2, p_value = chi2_contingency([observed, expected], correction=False)[:2] \n",
    "        \n",
    "        is_in_equilibrium = p_value >= alpha\n",
    "        \n",
    "        hwe_results[snp] = (chi2, p_value, is_in_equilibrium)\n",
    "    \n",
    "    return hwe_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a797b5d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_from_hwe_B(dataframe, hwe_results):\n",
    "\n",
    "    snps_to_keep = [snp for snp, (_, _, is_in_equilibrium) in hwe_results.items() if is_in_equilibrium]\n",
    "    filtered_dataframe = dataframe[snps_to_keep]\n",
    "    \n",
    "    return filtered_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "89e7e32d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 499504)\n"
     ]
    }
   ],
   "source": [
    "hwe_results = filter_snps_by_hwe_B(result_df, alpha=0.01)\n",
    "result_df = filter_snps_from_hwe_B(result_df, hwe_results)\n",
    "print(result_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82d88d16",
   "metadata": {},
   "source": [
    "### Variance Threshold Filtering "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca6c80b",
   "metadata": {},
   "source": [
    "Variance Threshold: Remove SNPs that have low variance across samples, as they might not be informative."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9db2a451",
   "metadata": {},
   "outputs": [],
   "source": [
    "def variance_threshold_filtering(df, p):\n",
    "    \n",
    "    threshold = p * (1 - p)\n",
    "    selector = VarianceThreshold(threshold)\n",
    "\n",
    "    df_transformed = selector.fit_transform(df)\n",
    "    df_transformed = pd.DataFrame(df_transformed, columns=df.columns[selector.get_support()], index= df.index)\n",
    "\n",
    "    kept_columns = df.columns[selector.get_support()]\n",
    "    dropped_columns = df.columns[~selector.get_support()]\n",
    "    \n",
    "    return df_transformed, kept_columns, dropped_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f26201a0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 477372)\n"
     ]
    }
   ],
   "source": [
    "result_df, _, _ = variance_threshold_filtering(result_df, 0.85)\n",
    "print(result_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "62fa50cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs12124819</th>\n",
       "      <th>rs6681049</th>\n",
       "      <th>rs4970383</th>\n",
       "      <th>rs4475691</th>\n",
       "      <th>rs7537756</th>\n",
       "      <th>rs3748597</th>\n",
       "      <th>rs28391282</th>\n",
       "      <th>rs2340592</th>\n",
       "      <th>rs1891910</th>\n",
       "      <th>rs3128117</th>\n",
       "      <th>...</th>\n",
       "      <th>rs739365</th>\n",
       "      <th>rs5770992</th>\n",
       "      <th>rs2040487</th>\n",
       "      <th>rs9628187</th>\n",
       "      <th>rs6010063</th>\n",
       "      <th>rs10451</th>\n",
       "      <th>rs715586</th>\n",
       "      <th>rs8137951</th>\n",
       "      <th>rs2301584</th>\n",
       "      <th>rs3810648</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Person_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 477372 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rs12124819  rs6681049  rs4970383  rs4475691  rs7537756  rs3748597  \\\n",
       "Person_ID                                                                      \n",
       "HG00096           0.0        2.0        0.0        0.0        0.0        2.0   \n",
       "HG00097           0.0        2.0        1.0        1.0        1.0        2.0   \n",
       "HG00099           0.0        2.0        0.0        0.0        0.0        2.0   \n",
       "\n",
       "           rs28391282  rs2340592  rs1891910  rs3128117  ...  rs739365  \\\n",
       "Person_ID                                               ...             \n",
       "HG00096           0.0        1.0        1.0        2.0  ...       0.0   \n",
       "HG00097           0.0        1.0        0.0        1.0  ...       0.0   \n",
       "HG00099           0.0        0.0        0.0        0.0  ...       0.0   \n",
       "\n",
       "           rs5770992  rs2040487  rs9628187  rs6010063  rs10451  rs715586  \\\n",
       "Person_ID                                                                  \n",
       "HG00096          0.0        2.0        0.0        1.0      0.0       1.0   \n",
       "HG00097          0.0        1.0        0.0        0.0      0.0       0.0   \n",
       "HG00099          0.0        1.0        0.0        1.0      1.0       0.0   \n",
       "\n",
       "           rs8137951  rs2301584  rs3810648  \n",
       "Person_ID                                   \n",
       "HG00096          0.0        0.0        0.0  \n",
       "HG00097          0.0        0.0        0.0  \n",
       "HG00099          1.0        0.0        0.0  \n",
       "\n",
       "[3 rows x 477372 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad16bb43",
   "metadata": {},
   "source": [
    "### Preparing Population Codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "52e51c24",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_labels = pop_labels[pop_labels['Sample name'].isin(result_df.index)]\n",
    "pop_labels = pop_labels[['Sample name', 'Superpopulation code']]\n",
    "pop_labels.rename(columns={\"Sample name\": \"Person_ID\", \"Superpopulation code\": \"labels\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79242141",
   "metadata": {},
   "source": [
    "### Principal Component Analysis (PCA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3a94a55d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_pca(df, n_components=None, explained_variance=None):\n",
    "\n",
    "    if n_components is not None:\n",
    "        pca = PCA(n_components=n_components)\n",
    "    elif explained_variance is not None:\n",
    "        pca = PCA(n_components=explained_variance)\n",
    "    else:\n",
    "        pca = PCA()\n",
    "    \n",
    "    principal_components = pca.fit_transform(df)\n",
    "    \n",
    "    col_names = [f'PC{i+1}' for i in range(principal_components.shape[1])]\n",
    "    pca_df = pd.DataFrame(data=principal_components, columns=col_names, index= df.index)\n",
    "    \n",
    "    return pca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "254fb7ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_data = apply_pca(result_df, n_components=1000, explained_variance=0.90)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bc6563ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "      <th>...</th>\n",
       "      <th>PC991</th>\n",
       "      <th>PC992</th>\n",
       "      <th>PC993</th>\n",
       "      <th>PC994</th>\n",
       "      <th>PC995</th>\n",
       "      <th>PC996</th>\n",
       "      <th>PC997</th>\n",
       "      <th>PC998</th>\n",
       "      <th>PC999</th>\n",
       "      <th>PC1000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Person_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>-38.639026</td>\n",
       "      <td>-90.794450</td>\n",
       "      <td>23.263929</td>\n",
       "      <td>-35.948275</td>\n",
       "      <td>1.161187</td>\n",
       "      <td>9.309609</td>\n",
       "      <td>-0.001401</td>\n",
       "      <td>2.061211</td>\n",
       "      <td>2.417782</td>\n",
       "      <td>6.942762</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.468540</td>\n",
       "      <td>-7.901774</td>\n",
       "      <td>9.770670</td>\n",
       "      <td>-11.258240</td>\n",
       "      <td>-7.951597</td>\n",
       "      <td>-2.844346</td>\n",
       "      <td>5.893723</td>\n",
       "      <td>2.953390</td>\n",
       "      <td>-4.820451</td>\n",
       "      <td>-3.457041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>-38.439010</td>\n",
       "      <td>-93.450054</td>\n",
       "      <td>22.201406</td>\n",
       "      <td>-36.684623</td>\n",
       "      <td>-0.139637</td>\n",
       "      <td>6.492132</td>\n",
       "      <td>-0.724865</td>\n",
       "      <td>-1.474823</td>\n",
       "      <td>0.837887</td>\n",
       "      <td>10.620088</td>\n",
       "      <td>...</td>\n",
       "      <td>3.764506</td>\n",
       "      <td>-4.264768</td>\n",
       "      <td>-13.994740</td>\n",
       "      <td>-14.554333</td>\n",
       "      <td>10.144423</td>\n",
       "      <td>-4.262681</td>\n",
       "      <td>-4.673796</td>\n",
       "      <td>11.506335</td>\n",
       "      <td>-5.590803</td>\n",
       "      <td>-9.564501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>-38.349643</td>\n",
       "      <td>-91.178250</td>\n",
       "      <td>24.402136</td>\n",
       "      <td>-35.676430</td>\n",
       "      <td>1.074924</td>\n",
       "      <td>14.443116</td>\n",
       "      <td>1.314256</td>\n",
       "      <td>-5.785633</td>\n",
       "      <td>8.926218</td>\n",
       "      <td>4.024488</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.241154</td>\n",
       "      <td>-11.742524</td>\n",
       "      <td>5.602238</td>\n",
       "      <td>2.376374</td>\n",
       "      <td>-14.044337</td>\n",
       "      <td>4.720400</td>\n",
       "      <td>3.871844</td>\n",
       "      <td>-0.231161</td>\n",
       "      <td>0.315130</td>\n",
       "      <td>-8.066297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PC1        PC2        PC3        PC4       PC5        PC6  \\\n",
       "Person_ID                                                                    \n",
       "HG00096   -38.639026 -90.794450  23.263929 -35.948275  1.161187   9.309609   \n",
       "HG00097   -38.439010 -93.450054  22.201406 -36.684623 -0.139637   6.492132   \n",
       "HG00099   -38.349643 -91.178250  24.402136 -35.676430  1.074924  14.443116   \n",
       "\n",
       "                PC7       PC8       PC9       PC10  ...     PC991      PC992  \\\n",
       "Person_ID                                           ...                        \n",
       "HG00096   -0.001401  2.061211  2.417782   6.942762  ... -5.468540  -7.901774   \n",
       "HG00097   -0.724865 -1.474823  0.837887  10.620088  ...  3.764506  -4.264768   \n",
       "HG00099    1.314256 -5.785633  8.926218   4.024488  ... -2.241154 -11.742524   \n",
       "\n",
       "               PC993      PC994      PC995     PC996     PC997      PC998  \\\n",
       "Person_ID                                                                   \n",
       "HG00096     9.770670 -11.258240  -7.951597 -2.844346  5.893723   2.953390   \n",
       "HG00097   -13.994740 -14.554333  10.144423 -4.262681 -4.673796  11.506335   \n",
       "HG00099     5.602238   2.376374 -14.044337  4.720400  3.871844  -0.231161   \n",
       "\n",
       "              PC999    PC1000  \n",
       "Person_ID                      \n",
       "HG00096   -4.820451 -3.457041  \n",
       "HG00097   -5.590803 -9.564501  \n",
       "HG00099    0.315130 -8.066297  \n",
       "\n",
       "[3 rows x 1000 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c51efddc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 1000)\n"
     ]
    }
   ],
   "source": [
    "print(pca_data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34c3c750",
   "metadata": {},
   "source": [
    "### Singular Value Decomposition (SVD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8d8f6e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_svd_sklearn(df, n_components=100):\n",
    "\n",
    "    svd = TruncatedSVD(n_components=n_components)\n",
    "    svd_fit = svd.fit_transform(df)\n",
    "    \n",
    "    col_names = [f'SVD{i+1}' for i in range(n_components)]\n",
    "    svd_df = pd.DataFrame(data=svd_fit, columns=col_names, index= df.index)\n",
    "    \n",
    "    return svd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e736a3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_svd_numpy(df, n_components=100):\n",
    "\n",
    "    centered_data = df - df.mean()\n",
    "    \n",
    "    U, s, Vt = np.linalg.svd(centered_data, full_matrices=False)  \n",
    "    \n",
    "    U_reduced  = U[:, :n_components]\n",
    "    s_reduced  = s[:n_components]\n",
    "    Vt_reduced = Vt[:n_components, :]\n",
    "\n",
    "    reduced_data = np.dot(U_reduced, np.diag(s_reduced))\n",
    "    \n",
    "    col_names = [f'SVD{i+1}' for i in range(n_components)]\n",
    "    svd_df = pd.DataFrame(data=reduced_data, columns=col_names, index= df.index)\n",
    "    \n",
    "    return svd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "37db9efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "svd_data = apply_svd_sklearn(result_df, n_components=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e62661da",
   "metadata": {},
   "outputs": [],
   "source": [
    "svd_data_df = apply_svd_numpy(result_df, n_components=1000)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5c7ffa36",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SVD1</th>\n",
       "      <th>SVD2</th>\n",
       "      <th>SVD3</th>\n",
       "      <th>SVD4</th>\n",
       "      <th>SVD5</th>\n",
       "      <th>SVD6</th>\n",
       "      <th>SVD7</th>\n",
       "      <th>SVD8</th>\n",
       "      <th>SVD9</th>\n",
       "      <th>SVD10</th>\n",
       "      <th>...</th>\n",
       "      <th>SVD991</th>\n",
       "      <th>SVD992</th>\n",
       "      <th>SVD993</th>\n",
       "      <th>SVD994</th>\n",
       "      <th>SVD995</th>\n",
       "      <th>SVD996</th>\n",
       "      <th>SVD997</th>\n",
       "      <th>SVD998</th>\n",
       "      <th>SVD999</th>\n",
       "      <th>SVD1000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Person_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>603.046621</td>\n",
       "      <td>-39.278318</td>\n",
       "      <td>-90.627046</td>\n",
       "      <td>23.277228</td>\n",
       "      <td>-36.099755</td>\n",
       "      <td>1.153453</td>\n",
       "      <td>9.316863</td>\n",
       "      <td>0.012396</td>\n",
       "      <td>2.155326</td>\n",
       "      <td>2.370304</td>\n",
       "      <td>...</td>\n",
       "      <td>3.156259</td>\n",
       "      <td>-1.474766</td>\n",
       "      <td>1.161947</td>\n",
       "      <td>0.780569</td>\n",
       "      <td>-1.222766</td>\n",
       "      <td>11.323725</td>\n",
       "      <td>-3.306233</td>\n",
       "      <td>11.323746</td>\n",
       "      <td>-2.614192</td>\n",
       "      <td>-3.830227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>601.038719</td>\n",
       "      <td>-39.184792</td>\n",
       "      <td>-93.150561</td>\n",
       "      <td>22.195952</td>\n",
       "      <td>-36.778369</td>\n",
       "      <td>-0.131148</td>\n",
       "      <td>6.513195</td>\n",
       "      <td>-0.761343</td>\n",
       "      <td>-1.570738</td>\n",
       "      <td>1.100527</td>\n",
       "      <td>...</td>\n",
       "      <td>7.461926</td>\n",
       "      <td>0.411498</td>\n",
       "      <td>-4.711832</td>\n",
       "      <td>1.720158</td>\n",
       "      <td>-6.858940</td>\n",
       "      <td>2.205314</td>\n",
       "      <td>-5.848622</td>\n",
       "      <td>-4.742616</td>\n",
       "      <td>-0.772183</td>\n",
       "      <td>7.926935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>605.540751</td>\n",
       "      <td>-38.887557</td>\n",
       "      <td>-91.187782</td>\n",
       "      <td>24.442167</td>\n",
       "      <td>-35.910246</td>\n",
       "      <td>1.037098</td>\n",
       "      <td>14.451912</td>\n",
       "      <td>1.300676</td>\n",
       "      <td>-6.059226</td>\n",
       "      <td>8.863826</td>\n",
       "      <td>...</td>\n",
       "      <td>2.322019</td>\n",
       "      <td>-10.555962</td>\n",
       "      <td>8.663292</td>\n",
       "      <td>6.594597</td>\n",
       "      <td>-6.732870</td>\n",
       "      <td>-12.189297</td>\n",
       "      <td>-2.027765</td>\n",
       "      <td>0.143378</td>\n",
       "      <td>-7.920362</td>\n",
       "      <td>2.626753</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SVD1       SVD2       SVD3       SVD4       SVD5      SVD6  \\\n",
       "Person_ID                                                                     \n",
       "HG00096    603.046621 -39.278318 -90.627046  23.277228 -36.099755  1.153453   \n",
       "HG00097    601.038719 -39.184792 -93.150561  22.195952 -36.778369 -0.131148   \n",
       "HG00099    605.540751 -38.887557 -91.187782  24.442167 -35.910246  1.037098   \n",
       "\n",
       "                SVD7      SVD8      SVD9     SVD10  ...    SVD991     SVD992  \\\n",
       "Person_ID                                           ...                        \n",
       "HG00096     9.316863  0.012396  2.155326  2.370304  ...  3.156259  -1.474766   \n",
       "HG00097     6.513195 -0.761343 -1.570738  1.100527  ...  7.461926   0.411498   \n",
       "HG00099    14.451912  1.300676 -6.059226  8.863826  ...  2.322019 -10.555962   \n",
       "\n",
       "             SVD993    SVD994    SVD995     SVD996    SVD997     SVD998  \\\n",
       "Person_ID                                                                 \n",
       "HG00096    1.161947  0.780569 -1.222766  11.323725 -3.306233  11.323746   \n",
       "HG00097   -4.711832  1.720158 -6.858940   2.205314 -5.848622  -4.742616   \n",
       "HG00099    8.663292  6.594597 -6.732870 -12.189297 -2.027765   0.143378   \n",
       "\n",
       "             SVD999   SVD1000  \n",
       "Person_ID                      \n",
       "HG00096   -2.614192 -3.830227  \n",
       "HG00097   -0.772183  7.926935  \n",
       "HG00099   -7.920362  2.626753  \n",
       "\n",
       "[3 rows x 1000 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svd_data.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1fa3db3",
   "metadata": {},
   "source": [
    "### Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b2e7f37a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_rf_feature_selection(df, labels_df, num_features=None, importance_threshold=None):\n",
    "    \n",
    "    merged_df = df.merge(labels_df, left_index=True, right_on='Person_ID')\n",
    "    \n",
    "    data = merged_df.drop(columns=['Person_ID', 'labels'])\n",
    "    labels = merged_df['labels']\n",
    "\n",
    "    rf = RandomForestClassifier(random_state=42)\n",
    "    rf.fit(data, labels)\n",
    "\n",
    "    feature_importances = rf.feature_importances_\n",
    "\n",
    "    if importance_threshold is not None:\n",
    "        features_to_keep = df.columns[feature_importances >= importance_threshold]\n",
    "    elif num_features is not None:\n",
    "        sorted_features = df.columns[np.argsort(-feature_importances)]\n",
    "        features_to_keep = sorted_features[:num_features]\n",
    "    else:\n",
    "        features_to_keep = df.columns\n",
    "\n",
    "    reduced_df = df[features_to_keep]\n",
    "    \n",
    "    return reduced_df, rf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d7281a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "reduced_df, rf = apply_rf_feature_selection(result_df, pop_labels, num_features=1000, importance_threshold=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4eb82f65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs7128655</th>\n",
       "      <th>rs1999047</th>\n",
       "      <th>rs10224196</th>\n",
       "      <th>rs1390151</th>\n",
       "      <th>rs1144628</th>\n",
       "      <th>rs17242334</th>\n",
       "      <th>rs11862829</th>\n",
       "      <th>rs12516794</th>\n",
       "      <th>rs6573875</th>\n",
       "      <th>rs2824119</th>\n",
       "      <th>...</th>\n",
       "      <th>rs11171949</th>\n",
       "      <th>rs10850526</th>\n",
       "      <th>rs2252523</th>\n",
       "      <th>rs9914327</th>\n",
       "      <th>rs8107859</th>\n",
       "      <th>rs12207656</th>\n",
       "      <th>rs7729096</th>\n",
       "      <th>rs7570835</th>\n",
       "      <th>rs577737</th>\n",
       "      <th>rs953062</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Person_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rs7128655  rs1999047  rs10224196  rs1390151  rs1144628  rs17242334  \\\n",
       "Person_ID                                                                       \n",
       "HG00096          0.0        1.0         0.0        0.0        1.0         0.0   \n",
       "HG00097          0.0        1.0         0.0        0.0        0.0         0.0   \n",
       "HG00099          0.0        1.0         0.0        0.0        1.0         0.0   \n",
       "\n",
       "           rs11862829  rs12516794  rs6573875  rs2824119  ...  rs11171949  \\\n",
       "Person_ID                                                ...               \n",
       "HG00096           1.0         1.0        1.0        0.0  ...         0.0   \n",
       "HG00097           2.0         0.0        2.0        1.0  ...         0.0   \n",
       "HG00099           2.0         0.0        1.0        0.0  ...         0.0   \n",
       "\n",
       "           rs10850526  rs2252523  rs9914327  rs8107859  rs12207656  rs7729096  \\\n",
       "Person_ID                                                                       \n",
       "HG00096           1.0        2.0        1.0        2.0         0.0        0.0   \n",
       "HG00097           1.0        2.0        0.0        2.0         0.0        0.0   \n",
       "HG00099           0.0        1.0        1.0        2.0         1.0        0.0   \n",
       "\n",
       "           rs7570835  rs577737  rs953062  \n",
       "Person_ID                                 \n",
       "HG00096          0.0       1.0       1.0  \n",
       "HG00097          0.0       1.0       1.0  \n",
       "HG00099          1.0       0.0       0.0  \n",
       "\n",
       "[3 rows x 1000 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reduced_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28978847",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c2e632a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0580a774",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "snp",
   "language": "python",
   "name": "snp"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
