{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f382f0a",
   "metadata": {},
   "source": [
    "### Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e13de0f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re  \n",
    "import fnmatch \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "\n",
    "from sklearn.ensemble import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "063e20b4",
   "metadata": {},
   "source": [
    "### Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e1a46a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "parquet_dir = '.../Data/parquet/'\n",
    "output_file = '.../Data/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f27bf68",
   "metadata": {},
   "source": [
    "### Read Population Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3c996649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123417</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00276</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123424</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00288</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME1839246</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample name     Sex Biosample ID Population code Population name  \\\n",
       "0     HG00271    male   SAME123417             FIN         Finnish   \n",
       "1     HG00276  female   SAME123424             FIN         Finnish   \n",
       "2     HG00288  female  SAME1839246             FIN         Finnish   \n",
       "\n",
       "  Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "0                  EUR    European Ancestry                   FIN   \n",
       "1                  EUR    European Ancestry                   FIN   \n",
       "2                  EUR    European Ancestry                   FIN   \n",
       "\n",
       "                                    Data collections  \n",
       "0  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "1  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_labels = pd.read_csv(parquet_dir + 'igsr_samples.tsv', sep='\\t')\n",
    "pop_labels.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35151422",
   "metadata": {},
   "source": [
    "### Read, Clean, and Merge SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa586ae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_and_clean_chromosome_files(folder_path):\n",
    "\n",
    "    for chr_num in range(1, 23):\n",
    "        \n",
    "        file_path = f\"{folder_path}/chr{chr_num}.parquet\"\n",
    "        df = pd.read_parquet(file_path, engine='pyarrow')\n",
    "        \n",
    "        missing_percentage = df.isnull().mean() * 100\n",
    "        cols_to_keep = missing_percentage[missing_percentage <= 5].index\n",
    "        cleaned_df = df[cols_to_keep]\n",
    "        \n",
    "        if chr_num == 1:\n",
    "            final = cleaned_df.copy()\n",
    "        else:\n",
    "            final = final.merge(cleaned_df, on='Person_ID')\n",
    "                \n",
    "    return final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "09a60b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = read_and_clean_chromosome_files(parquet_dir)\n",
    "result_df = result_df.set_index('Person_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eceb4311",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs3094315</th>\n",
       "      <th>rs3131972</th>\n",
       "      <th>rs12124819</th>\n",
       "      <th>rs11240777</th>\n",
       "      <th>rs6681049</th>\n",
       "      <th>rs4970383</th>\n",
       "      <th>rs4475691</th>\n",
       "      <th>rs7537756</th>\n",
       "      <th>rs13302982</th>\n",
       "      <th>rs1110052</th>\n",
       "      <th>...</th>\n",
       "      <th>rs10451</th>\n",
       "      <th>rs715586</th>\n",
       "      <th>rs8137951</th>\n",
       "      <th>rs2301584</th>\n",
       "      <th>rs756638</th>\n",
       "      <th>rs3810648</th>\n",
       "      <th>rs2285395</th>\n",
       "      <th>rs13056621</th>\n",
       "      <th>rs5771007</th>\n",
       "      <th>rs3888396</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Person_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 909218 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rs3094315  rs3131972  rs12124819  rs11240777  rs6681049  rs4970383  \\\n",
       "Person_ID                                                                       \n",
       "HG00096            1          1           0           1          2          0   \n",
       "HG00097            2          2           0           0          2          1   \n",
       "HG00099            1          1           0           1          2          0   \n",
       "\n",
       "           rs4475691  rs7537756  rs13302982  rs1110052  ...  rs10451  \\\n",
       "Person_ID                                               ...            \n",
       "HG00096            0          0           2          1  ...        0   \n",
       "HG00097            1          1           2          0  ...        0   \n",
       "HG00099            0          0           2          2  ...        1   \n",
       "\n",
       "           rs715586  rs8137951  rs2301584  rs756638  rs3810648  rs2285395  \\\n",
       "Person_ID                                                                   \n",
       "HG00096           1          0          0         0          0          0   \n",
       "HG00097           0          0          0         0          0          0   \n",
       "HG00099           0          1          0         0          0          0   \n",
       "\n",
       "           rs13056621  rs5771007  rs3888396  \n",
       "Person_ID                                    \n",
       "HG00096             1          0          1  \n",
       "HG00097             1          0          0  \n",
       "HG00099             2          0          0  \n",
       "\n",
       "[3 rows x 909218 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8a9a2c",
   "metadata": {},
   "source": [
    "### Outlier Detection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c3f3949",
   "metadata": {},
   "source": [
    "Identify all values that are outside the range of 0 to 2 (inclusive), and set those values to NaN, which represents missing data in pandas DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b14555e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_snp_data(df):\n",
    "    \n",
    "    mask = (df < 0) | (df > 2)\n",
    "    df[mask] = np.nan\n",
    "    \n",
    "    return df\n",
    "\n",
    "result_df = clean_snp_data(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ceb30e",
   "metadata": {},
   "source": [
    "### Remove SNPs with Missing Values > 5% "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc207cd6",
   "metadata": {},
   "source": [
    "Remove: If a sample (row) or SNP (column) has a high percentage of missing values, you might consider removing it entirely from the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a492bf22",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 0.05 * result_df.shape[0]  \n",
    "result_df = result_df.dropna(axis='columns', thresh=threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8318f217",
   "metadata": {},
   "source": [
    "### Impute Missing Values with Mode"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7112e43",
   "metadata": {},
   "source": [
    "Impute: For SNPs with a small percentage of missing data, you can impute missing values. Common strategies include replacing missing values with the median or mode. In SNP data, using the most common genotype (0, 1, or 2) among the observed values for that SNP can be a sensible approach."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7460e1ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def selective_imputation(df):\n",
    "    \n",
    "    missing_columns = df.columns[df.isnull().any()].tolist()\n",
    "\n",
    "    imputer = SimpleImputer(strategy='most_frequent')\n",
    "    \n",
    "    if missing_columns:\n",
    "        df[missing_columns] = imputer.fit_transform(df[missing_columns])\n",
    "    \n",
    "    return df\n",
    "\n",
    "result_df = selective_imputation(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b236c0c",
   "metadata": {},
   "source": [
    "### Minor Allele Frequency (MAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61c7ea58",
   "metadata": {},
   "source": [
    "SNPs with a very low MAF are often excluded since they may not contribute much information for ancestry prediction and can lead to model overfitting. A common threshold is to exclude SNPs with a MAF less than 5%."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bd9988a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_maf_A(df, maf_threshold=0.05):\n",
    "\n",
    "    maf_dict = {}\n",
    "    total_alleles = 2 * len(df)\n",
    "    \n",
    "    for column in df.columns:\n",
    "        \n",
    "        counts = df[column].value_counts()\n",
    "        \n",
    "        freq_minor_allele = min((2 * counts.get(2, 0) + counts.get(1, 0)) / total_alleles,\n",
    "                                (2 * counts.get(0, 0) + counts.get(1, 0)) / total_alleles)\n",
    "        \n",
    "        maf_dict[column] = freq_minor_allele\n",
    "    \n",
    "    columns_to_keep = [column for column, maf in maf_dict.items() if maf >= maf_threshold]\n",
    "    \n",
    "    filtered_df = df[columns_to_keep]\n",
    "    \n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f3b19a2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_maf_B(dataframe, maf_threshold=0.05):\n",
    "\n",
    "    maf_dict = {}\n",
    "    to_drop  = []\n",
    "\n",
    "    for snp in dataframe.columns:\n",
    "\n",
    "        genotype_counts = dataframe[snp].value_counts()\n",
    "        \n",
    "        n_AA = genotype_counts.get(0, 0)\n",
    "        n_Aa = genotype_counts.get(1, 0)\n",
    "        n_aa = genotype_counts.get(2, 0)\n",
    "        \n",
    "        total_alleles = 2 * (n_AA + n_Aa + n_aa)\n",
    "        freq_A = (2 * n_AA + n_Aa) / total_alleles\n",
    "        freq_a = 1 - freq_A  \n",
    "        \n",
    "        maf = min(freq_A, freq_a)\n",
    "        maf_dict[snp] = maf\n",
    "        \n",
    "        if maf < maf_threshold:\n",
    "            to_drop.append(snp)\n",
    "    \n",
    "    filtered_df = dataframe.drop(columns=to_drop)\n",
    "\n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "512d0566",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 823111)\n"
     ]
    }
   ],
   "source": [
    "result_df = filter_snps_by_maf_B(result_df, maf_threshold=0.05)\n",
    "print(result_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e47cd792",
   "metadata": {},
   "source": [
    "### Hardy-Weinberg Equilibrium (HWE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9255ee00",
   "metadata": {},
   "source": [
    "Deviations from HWE can indicate problems such as genotyping errors. While checking for HWE is more common in association studies, you might decide to perform this check and exclude SNPs not in equilibrium if you suspect data quality issues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cb036515",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_hwe_A(df, hwe_p_threshold=0.01):\n",
    "\n",
    "    columns_to_keep = []\n",
    "    \n",
    "    for column in df.columns:\n",
    "        \n",
    "        genotype_counts = df[column].value_counts().sort_index()\n",
    "\n",
    "        for genotype in [0, 1, 2]:  \n",
    "            if genotype not in genotype_counts.index:\n",
    "                genotype_counts.at[genotype] = 0\n",
    "\n",
    "        p = (2 * genotype_counts[0] + genotype_counts[1]) / (2 * sum(genotype_counts.values)) \n",
    "        q = 1 - p \n",
    "\n",
    "        expected_counts = pd.Series({0: (p ** 2)  * sum(genotype_counts.values),  \n",
    "                                     1: 2 * p * q * sum(genotype_counts.values), \n",
    "                                     2: (q ** 2)  * sum(genotype_counts.values)})\n",
    "\n",
    "        chi2, p_value = stats.chisquare(f_obs=genotype_counts, f_exp=expected_counts)\n",
    "\n",
    "        if p_value >= hwe_p_threshold:\n",
    "            columns_to_keep.append(column)\n",
    "            \n",
    "    filtered_df = df[columns_to_keep]\n",
    "    \n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a3bf51fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_hwe_B(dataframe, alpha=0.01):\n",
    "\n",
    "    hwe_results = {}\n",
    "    \n",
    "    for snp in dataframe.columns:\n",
    "        \n",
    "        genotype_counts = dataframe[snp].value_counts()\n",
    "        \n",
    "        obs_AA = genotype_counts.get(0, 0)\n",
    "        obs_Aa = genotype_counts.get(1, 0)\n",
    "        obs_aa = genotype_counts.get(2, 0)\n",
    "        \n",
    "        total_alleles = 2 * (obs_AA + obs_Aa + obs_aa)\n",
    "        p = (2 * obs_AA + obs_Aa) / total_alleles \n",
    "        q = 1 - p  \n",
    "        \n",
    "        exp_AA = (p ** 2)  * (obs_AA + obs_Aa + obs_aa)\n",
    "        exp_Aa = 2 * p * q * (obs_AA + obs_Aa + obs_aa)\n",
    "        exp_aa = (q ** 2)  * (obs_AA + obs_Aa + obs_aa)\n",
    "        \n",
    "        observed = np.array([obs_AA, obs_Aa, obs_aa])\n",
    "        expected = np.array([exp_AA, exp_Aa, exp_aa])\n",
    "        chi2, p_value = chi2_contingency([observed, expected], correction=False)[:2] \n",
    "        \n",
    "        is_in_equilibrium = p_value >= alpha\n",
    "        \n",
    "        hwe_results[snp] = (chi2, p_value, is_in_equilibrium)\n",
    "    \n",
    "    return hwe_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a797b5d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_from_hwe_B(dataframe, hwe_results):\n",
    "\n",
    "    snps_to_keep = [snp for snp, (_, _, is_in_equilibrium) in hwe_results.items() if is_in_equilibrium]\n",
    "    filtered_dataframe = dataframe[snps_to_keep]\n",
    "    \n",
    "    return filtered_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "89e7e32d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 499504)\n"
     ]
    }
   ],
   "source": [
    "hwe_results = filter_snps_by_hwe_B(result_df, alpha=0.01)\n",
    "result_df = filter_snps_from_hwe_B(result_df, hwe_results)\n",
    "print(result_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82d88d16",
   "metadata": {},
   "source": [
    "### Variance Threshold Filtering "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca6c80b",
   "metadata": {},
   "source": [
    "Variance Threshold: Remove SNPs that have low variance across samples, as they might not be informative."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9db2a451",
   "metadata": {},
   "outputs": [],
   "source": [
    "def variance_threshold_filtering(df, p):\n",
    "    \n",
    "    threshold = p * (1 - p)\n",
    "    selector = VarianceThreshold(threshold)\n",
    "\n",
    "    df_transformed = selector.fit_transform(df)\n",
    "    df_transformed = pd.DataFrame(df_transformed, columns=df.columns[selector.get_support()], index= df.index)\n",
    "\n",
    "    kept_columns = df.columns[selector.get_support()]\n",
    "    dropped_columns = df.columns[~selector.get_support()]\n",
    "    \n",
    "    return df_transformed, kept_columns, dropped_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f26201a0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2504, 477372)\n"
     ]
    }
   ],
   "source": [
    "result_df, _, _ = variance_threshold_filtering(result_df, 0.85)\n",
    "print(result_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "62fa50cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs12124819</th>\n",
       "      <th>rs6681049</th>\n",
       "      <th>rs4970383</th>\n",
       "      <th>rs4475691</th>\n",
       "      <th>rs7537756</th>\n",
       "      <th>rs3748597</th>\n",
       "      <th>rs28391282</th>\n",
       "      <th>rs2340592</th>\n",
       "      <th>rs1891910</th>\n",
       "      <th>rs3128117</th>\n",
       "      <th>...</th>\n",
       "      <th>rs739365</th>\n",
       "      <th>rs5770992</th>\n",
       "      <th>rs2040487</th>\n",
       "      <th>rs9628187</th>\n",
       "      <th>rs6010063</th>\n",
       "      <th>rs10451</th>\n",
       "      <th>rs715586</th>\n",
       "      <th>rs8137951</th>\n",
       "      <th>rs2301584</th>\n",
       "      <th>rs3810648</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Person_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HG00096</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00097</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG00099</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 477372 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rs12124819  rs6681049  rs4970383  rs4475691  rs7537756  rs3748597  \\\n",
       "Person_ID                                                                      \n",
       "HG00096           0.0        2.0        0.0        0.0        0.0        2.0   \n",
       "HG00097           0.0        2.0        1.0        1.0        1.0        2.0   \n",
       "HG00099           0.0        2.0        0.0        0.0        0.0        2.0   \n",
       "\n",
       "           rs28391282  rs2340592  rs1891910  rs3128117  ...  rs739365  \\\n",
       "Person_ID                                               ...             \n",
       "HG00096           0.0        1.0        1.0        2.0  ...       0.0   \n",
       "HG00097           0.0        1.0        0.0        1.0  ...       0.0   \n",
       "HG00099           0.0        0.0        0.0        0.0  ...       0.0   \n",
       "\n",
       "           rs5770992  rs2040487  rs9628187  rs6010063  rs10451  rs715586  \\\n",
       "Person_ID                                                                  \n",
       "HG00096          0.0        2.0        0.0        1.0      0.0       1.0   \n",
       "HG00097          0.0        1.0        0.0        0.0      0.0       0.0   \n",
       "HG00099          0.0        1.0        0.0        1.0      1.0       0.0   \n",
       "\n",
       "           rs8137951  rs2301584  rs3810648  \n",
       "Person_ID                                   \n",
       "HG00096          0.0        0.0        0.0  \n",
       "HG00097          0.0        0.0        0.0  \n",
       "HG00099          1.0        0.0        0.0  \n",
       "\n",
       "[3 rows x 477372 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "320d0115",
   "metadata": {},
   "source": [
    "### Save Clean Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "358378a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Person_ID</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>HG00103</th>\n",
       "      <th>HG00105</th>\n",
       "      <th>HG00106</th>\n",
       "      <th>HG00107</th>\n",
       "      <th>...</th>\n",
       "      <th>NA21128</th>\n",
       "      <th>NA21129</th>\n",
       "      <th>NA21130</th>\n",
       "      <th>NA21133</th>\n",
       "      <th>NA21135</th>\n",
       "      <th>NA21137</th>\n",
       "      <th>NA21141</th>\n",
       "      <th>NA21142</th>\n",
       "      <th>NA21143</th>\n",
       "      <th>NA21144</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs12124819</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs6681049</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4970383</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 2504 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Person_ID   HG00096  HG00097  HG00099  HG00100  HG00101  HG00102  HG00103  \\\n",
       "rs12124819      0.0      0.0      0.0      1.0      0.0      1.0      1.0   \n",
       "rs6681049       2.0      2.0      2.0      2.0      2.0      2.0      2.0   \n",
       "rs4970383       0.0      1.0      0.0      2.0      1.0      0.0      0.0   \n",
       "\n",
       "Person_ID   HG00105  HG00106  HG00107  ...  NA21128  NA21129  NA21130  \\\n",
       "rs12124819      1.0      1.0      1.0  ...      0.0      0.0      0.0   \n",
       "rs6681049       2.0      2.0      2.0  ...      2.0      2.0      2.0   \n",
       "rs4970383       0.0      0.0      1.0  ...      0.0      0.0      2.0   \n",
       "\n",
       "Person_ID   NA21133  NA21135  NA21137  NA21141  NA21142  NA21143  NA21144  \n",
       "rs12124819      0.0      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "rs6681049       2.0      2.0      2.0      2.0      2.0      2.0      2.0  \n",
       "rs4970383       2.0      0.0      1.0      1.0      1.0      1.0      0.0  \n",
       "\n",
       "[3 rows x 2504 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transposed_df = result_df.transpose()\n",
    "transposed_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "401d1cab",
   "metadata": {},
   "outputs": [],
   "source": [
    "table = pa.Table.from_pandas(transposed_df)\n",
    "pq.write_table(table, parquet_dir + \"Final_Data.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0580a774",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "snp",
   "language": "python",
   "name": "snp"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
